### PR TITLE
[in_app_purchase] migrate playing billing library to v3

### PR DESCRIPTION
**Pull request by cyanglaz**
This PR migrates the play billing library to V3. Several breaking changes are introduced due to the breaking changes in PBL v3. 

Fixes https://github.com/flutter/flutter/issues/75204

The release note can be found here, which can be used as a reference for updated APIs in this PR. https://developer.android.com/google/play/billing/release-notes#3-0

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/in_app_purchase/android/build.gradle [hunk 1]
```diff
@@ -35,9 +35,9 @@ android {
 
 dependencies {
     implementation 'androidx.annotation:annotation:1.0.0'
-    implementation 'com.android.billingclient:billing:2.0.3'
+    implementation 'com.android.billingclient:billing:3.0.2'
     testImplementation 'junit:junit:4.12'
-    testImplementation 'org.mockito:mockito-core:2.17.0'
+    testImplementation 'org.mockito:mockito-core:3.6.0'
```
Comment (cyanglaz) [inline packages/in_app_purchase/android/build.gradle:40]: This is required to mock the objects in the new PBL.

File: packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java [hunk 2]
```diff
@@ -286,7 +289,6 @@ public void launchBillingFlow_ok_null_AccountId() {
     verify(mockBillingClient).launchBillingFlow(any(), billingFlowParamsCaptor.capture());
     BillingFlowParams params = billingFlowParamsCaptor.getValue();
     assertEquals(params.getSku(), skuId);
-    assertNull(params.getAccountId());
```
Comment (cyanglaz) [inline packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:289]: getAccountId is not available anymore so we can't really test this value, however, we can still test if the call crashes, so ill leave the test here

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 3]
```diff
@@ -155,8 +155,13 @@ class BillingClient {
   /// The [skuDetails] needs to have already been fetched in a [querySkuDetails]
   /// call. The [accountId] is an optional hashed string associated with the user
   /// that's unique to your app. It's used by Google to detect unusual behavior.
-  /// Do not pass in a cleartext [accountId], use your developer ID, or use the
-  /// user's Google ID for this field.
+  /// Do not pass in a cleartext [accountId], Do not use this field to store any Personally Identifiable Information (PII)
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:158]: Nit: s/Do not use/and do not use/

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 4]
```diff
@@ -155,8 +155,13 @@ class BillingClient {
...
+  /// such as emails in cleartext. Attempting to store PII in this field will result in purchases being blocked.
+  /// Google Play recommends that you use either encryption or a one-way hash to generate an obfuscated identifier to send to Google Play.
+  ///
+  /// Specifies an optional [obfuscatedProfileId]  that is uniquely associated with the user's profile in your app.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:162]: Nit: Fix double space after ]

File: packages/in_app_purchase/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,12 @@
+## 0.5.0
+
+* Migrate to Google Billing Library 3.0
+  * Add `obfuscatedProfileId`, `purchaseToken` in [BillingClientWrapper.launchBillingFlow].
+  * Removed `developerPayload` in [BillingClientWrapper.acknowledgePurchase], [BillingClientWrapper.consumeAsync], [InAppPurchaseConnection.completePurchase], [InAppPurchaseConnection.consumePurchase].
```
Comment (stuartmorgan) [inline packages/in_app_purchase/CHANGELOG.md:5]: These aren't breaking changes?

File: packages/in_app_purchase/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,12 @@
...
+  * **Breaking Change**
+    * Removed `isRewarded` from [SkuDetailsWrapper].
```
Comment (stuartmorgan) [inline packages/in_app_purchase/CHANGELOG.md:7]: Is there any useful context to provide about why they are removed? E.g., link to discussion of removal in the SDK?

File: packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java (hunk 2 above)
Comment (stuartmorgan) [reply packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:289]: Can you rename the test then, to better explain what it's doing now? (Probably needs a comment in addition to a new name.)

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 7]
```diff
@@ -171,25 +176,31 @@ class BillingClient {
   /// [`BillingFlowParams`](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams)
   /// instance by [setting the given
   /// skuDetails](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder.html#setskudetails)
-  /// and [the given
-  /// accountId](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder.html#setAccountId(java.lang.String)).
+  /// , [the given
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:179]: Nit: this is hard to read; let's put the comma on the previous line.

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 8]
```diff
@@ -171,25 +176,31 @@ class BillingClient {
...
+  /// accountId](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder#setObfuscatedAccountId(java.lang.String))
+  /// and the [obfuscatedProfileId] (https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder#setobfuscatedprofileid).
   ///
   /// When this method is called to purchase a subscription, an optional `oldSku`
   /// can be passed in. This will tell Google Play that rather than purchasing a new subscription,
   /// the user needs to upgrade/downgrade the existing subscription.
-  /// The [oldSku](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder#setoldsku) is the SKU id that the user is upgrading or downgrading from.
+  /// The [oldSku](https://developer.android.com/reference/com/android/billingclient/api/BillingFlowParams.Builder#setoldsku) and [purchaseToken] are the SKU id and purchase token that the user is upgrading or downgrading from.
+  /// [purchaseToken] must not be `null` if [oldSku] is not `null`.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:187]: This should be asserted.

File: packages/in_app_purchase/lib/src/billing_client_wrappers/purchase_wrapper.dart [hunk 9]
```diff
@@ -120,6 +120,8 @@ class PurchaseWrapper {
   /// The payload specified by the developer when the purchase was acknowledged or consumed.
   ///
   /// The value is `null` if it wasn't specified when the purchase was acknowledged or consumed.
+  /// The `developerPayload` is removed from [BillingClientWrapper.acknowledgePurchase], [BillingClientWrapper.consumeAsync], [InAppPurchaseConnection.completePurchase], [InAppPurchaseConnection.consumePurchase]
+  /// after plugin version `0.5.0`. As a result, this will be `null` for new purchases happened after `0.5.0`.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/purchase_wrapper.dart:124]: s/happened after/that happen after updating to/

File: packages/in_app_purchase/lib/src/billing_client_wrappers/sku_details_wrapper.dart [hunk 10]
```diff
@@ -71,9 +70,11 @@ class SkuDetailsWrapper {
   @JsonKey(defaultValue: '')
   final String introductoryPriceMicros;
 
-  /// The number of billing perios that [introductoryPrice] is valid for ("2").
-  @JsonKey(defaultValue: '')
-  final String introductoryPriceCycles;
+  /// The number of subscription billing periods for which the user will be given the introductory price, such as 3.
+  /// Returns 0 if the SKU is not a subscription or doesn't have an introductory period.
+  /// Returns -1 if the value is not available.
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/sku_details_wrapper.dart:75]: What does "not available" mean as distinct from "doesn't have an introductory period"?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
We don't need any new tests to cover new functionality related to the added arguments?

File: packages/in_app_purchase/CHANGELOG.md (hunk 5 above)
Comment (cyanglaz) [reply packages/in_app_purchase/CHANGELOG.md]: Oh they are. Fixed

File: packages/in_app_purchase/CHANGELOG.md (hunk 6 above)
Comment (cyanglaz) [reply packages/in_app_purchase/CHANGELOG.md:7]: added

File: packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java (hunk 2 above)
Comment (cyanglaz) [reply packages/in_app_purchase/example/android/app/src/test/java/io/flutter/plugins/inapppurchase/MethodCallHandlerTest.java:289]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart (hunk 3 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart (hunk 4 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart (hunk 7 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart (hunk 8 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:185]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/purchase_wrapper.dart (hunk 9 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/purchase_wrapper.dart]: done

File: packages/in_app_purchase/lib/src/billing_client_wrappers/sku_details_wrapper.dart (hunk 10 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/billing_client_wrappers/sku_details_wrapper.dart]: Should be the same. Fixed.

**Review by cyanglaz — COMMENTED** [review]
@stuartmorgan Updated per your comments PTAL

File: packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart [hunk 11]
```diff
@@ -195,6 +193,8 @@ class BillingClient {
       String? purchaseToken,
       ProrationMode? prorationMode}) async {
     assert(sku != null);
+    assert((oldSku == null) == (purchaseToken == null),
+        'oldSku and purchaseToken should have the same nullability');
```
Comment (stuartmorgan) [inline packages/in_app_purchase/lib/src/billing_client_wrappers/billing_client_wrapper.dart:197]: Nit: nullability is a property of a type, not a value; the two types do have the same nullability (both allow null). s/should have the same nullability/must both be set, or both be null/

**Review by stuartmorgan — APPROVED** [review]
LGTM with final nit.

**Pull request merged**